### PR TITLE
Revert: Revert "Set val escape on expression variables (#64414)"

### PR DESCRIPTION
**Pull request by cston**
Commit 1: Revert: Revert "Set val escape on expression variables (#64414)"
Commit 2: Update escape analysis to ignore local declarations only.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs [hunk 1]
```diff
@@ -648,7 +648,13 @@ private string GetDebuggerDisplay()
                 {
                     var variableOpt = variablesOpt?[i].Single;
                     uint valEscape = variableOpt is null ? LocalScopeDepth : GetValEscape(variableOpt, LocalScopeDepth);
-                    var variable = new OutDeconstructVarPendingInference(receiverSyntax, valEscape);
+                    var variableSymbol = variableOpt switch
+                    {
+                        DeconstructionVariablePendingInference { VariableSymbol: var symbol } => symbol,
+                        BoundLocal { DeclarationKind: BoundLocalDeclarationKind.WithExplicitType or BoundLocalDeclarationKind.WithInferredType, LocalSymbol: var symbol } => symbol,
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:654]: We have other cases where we check `DeclarationKind: not BoundLocalDeclarationKind.None`. In future we might want to be consistent here, or maybe introduce an extension/computed property which asks if the bound-local represents a user-visible variable declaration, which throws when an unknown DeclarationKind is encountered.

No need to change anything in this PR.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 2]
```diff
@@ -8363,33 +8350,12 @@ static R2 F4(ref int i)
...
-                    //         var (x3, y3) = r;
-                    Diagnostic(ErrorCode.ERR_CallArgMixing, "r").WithArguments("R1.Deconstruct(out int, out R2)", "this").WithLocation(28, 24),
-                    // (34,9): error CS8352: Cannot use variable 'r' in this context because it may expose referenced variables outside of their declaration scope
-                    //         r.Deconstruct(out var x4, out var y4);
-                    Diagnostic(ErrorCode.ERR_EscapeVariable, "r").WithArguments("r").WithLocation(34, 9),
-                    // (34,9): error CS8350: This combination of arguments to 'R1.Deconstruct(out int, out R2)' is disallowed because it may expose variables referenced by parameter 'this' outside of their declaration scope
-                    //         r.Deconstruct(out var x4, out var y4);
-                    Diagnostic(ErrorCode.ERR_CallArgMixing, "r.Deconstruct(out var x4, out var y4)").WithArguments("R1.Deconstruct(out int, out R2)", "this").WithLocation(34, 9));
+                // (16,16): error CS8352: Cannot use variable 'y1' in this context because it may expose referenced variables outside of their declaration scope
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:8353]: nit: consider indenting this to match the previous test baseline

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:8353]: I'll update the indenting in a follow-up PR, unless other changes are needed in this PR.

**Comment by RikkiGibson** [comment]
The [second commit](https://github.com/dotnet/roslyn/pull/64633/commits/a4f21f09ec5e0642830e02d4be973310752e296d) needs a second review.

**Review by jaredpar — APPROVED** [review]

**Pull request merged**